packages/ens-utils/src/ethregistrar.ts

```diff
@@ -3,323 +3,351 @@ import {
+        // TODO: check for possible off-by-1 errors in the logic above
+        return false;
       }
     }
 
-    // TODO: refine this. For example, there's a maximum length limit, etc.
-    return true;
-  }
-}
+    if (duration.seconds < 1n) {
+      // TODO: enforce that `duration` is the minimum duration allowed for a registration
+      // TODO: need to put the right constant here.
+      return false;
+    }
 
-export class ClassicETHRegistrarController implements Registrar {
-  constructor(public contract: ContractRef) {
-    this.contract = contract;
+    return true;
   }
 
-  getTokenId(name: ENSName, isWrapped: boolean): TokenId {
-    if (!this.isClaimable(name, isWrapped)) {
+  public getRegistrationPriceQuote(
+    name: ENSName,
+    atTimestamp: Timestamp,
+    duration: Duration,
+    existingRegistration?: Registration,
+  ): RegistrationPriceQuote {
+    if (!this.canRegister(name, atTimestamp, duration, existingRegistration)) {
+      // TODO: refine the way we handle these exception cases.
       throw new Error(
-        `Unwrapped tokenId for name: "${name.name}" is not claimable by registrar: ${this.contract.address} on chainId: ${this.contract.chain.chainId}`,
+        `Cannot build registration price quote for name: "${name.name}".`,
+      );
+    }
+
+    const rentalPeriod = buildTimePeriod(
+      atTimestamp,
+      addSeconds(atTimestamp, duration),
+    );
+
+    let charges = this.getDurationCharges(name, duration);
+
+    let temporaryPremium: RegistrarTemporaryFee | null = null;
+
+    if (existingRegistration) {
+      temporaryPremium = this.getTemporaryPremiumCharge(
+        name,
+        existingRegistration,
+        atTimestamp,
       );
+      if (temporaryPremium) {
+        charges = [...charges, temporaryPremium];
+      }
     }
-    return buildTokenId(BigInt(labelhash(name.labels[0])));
+
+    const totalPrice = addPrices(charges.map((charge) => charge.price));
+
+    return {
+      action: RegistrarAction.Register,
+      rentalPeriod,
+      charges,
+      totalPrice,
+    };
   }
 
-  isClaimable(name: ENSName, isWrapped: boolean): boolean {
-    // name must be unwrapped
-    if (isWrapped) return false;
+  public canRenew(
+    name: ENSName,
+    atTimestamp: Timestamp,
+    duration: Duration,
+    existingRegistration: Registration,
+  ): boolean {
+    if (!this.getManagedSubname(name)) {
+      // name is not directly managed by this registrar
+      return false;
+    }
 
-    // must have exactly 2 labels to be a direct subname of ".eth"
-    if (name.labels.length !== 2) return false;
+    if (existingRegistration) {
+      if (
+        existingRegistration.registrationTimestamp &&
+        atTimestamp.time < existingRegistration.registrationTimestamp.time
+      ) {
+        // if the renewal is requested before the registration, we can't renew it
+        // TODO: check for possible off-by-1 errors in the logic above
```

Comment:
@lightwalker-eth can you extend on off-by-1 spec?

Reply:
We need to see how ENS smart contracts actually work. Ex: Actual expiration time? 1 second earlier? 1 second later?

Reply:
OK...

I'd say there is no problem with current logic.

If `atTimestamp` is the same than `registrationTimestamp`, it is already possible to renew it for longer periods, meaning `canRenew` should go on in the function scope, but, if it is smaller, it really is not possible to renew this domain: as per my review this has achieved its goal.